homeworks/07-mars-viewer/src/features/mars/marsSlice.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  createAsyncThunk,
+  createSelector,
+  createSlice,
+  PayloadAction,
+} from "@reduxjs/toolkit";
+import { RootState } from "../../app/store";
+import api from "../../api";
+import { MarsState, Rover, Sol } from "./types";
+
+const initialState: MarsState = {
+  rover: {
+    perseverance: {
+      photos: [],
+      sols: {},
+      status: "idle",
+      selectedSol: 1,
+    },
+  },
+  selectedRover: "perseverance",
+};
+
+export const getPhotos = createAsyncThunk<Sol[], void, { state: RootState }>(
+  "sols/getSols",
+  async (_, { getState }) => {
+    const {
+      mars: { rover, selectedRover },
+    } = getState();
+    const response = await api.getPhotos(
+      selectedRover,
+      rover[selectedRover].selectedSol
+    );
+    return response.photos;
+  }
+);
+
+export const marsSlice = createSlice({
+  name: "sols",
+  initialState,
+  reducers: {
+    setSol: (state, action: PayloadAction<number>) => {
+      state.rover[state.selectedRover].selectedSol = action.payload;
+    },
+    setRover: (state, action: PayloadAction<Rover>) => {
+      state.selectedRover = action.payload;
+      if (!state.rover[action.payload]) {
+        state.rover[action.payload] = {
+          photos: [],
+          sols: {},
+          status: "idle",
+          selectedSol: 1,
+        };
+      }
+    },
+  },
+  extraReducers: (builder) => {
+    builder
+      .addCase(getPhotos.pending, (state) => {
+        state.rover[state.selectedRover].status = "loading";
+      })
+      .addCase(getPhotos.fulfilled, (state, action) => {
+        state.rover[state.selectedRover].status = "idle";
+        if (!state.rover[state.selectedRover].photos) {
+          state.rover[state.selectedRover].photos = [];
+        }
+        state.rover[state.selectedRover].photos?.push(...action.payload);
```

Comment:
pushing without checking for duplicates leads to bug: if load the same day multiple times, there would be multiple copies of one picture and in favourites tab whole array being filtered and favourite image shows multiple times

Reply:
Bun in this case it wouldn't be possible, because clicking on the heart again removes an item from favorites.